### PR TITLE
stlink-tool: init at 0-unstable-2020-06-10

### PR DESCRIPTION
**Pull request by wucke13**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wucke13** [comment]
@GrahamcOfBorg build stlink-tool

**Comment by wucke13** [comment]
@GrahamcOfBorg build stlink-tool

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -25389,6 +25389,8 @@ with pkgs;
   stlink = callPackage ../development/tools/misc/stlink { };
   stlink-gui = callPackage ../development/tools/misc/stlink { withGUI = true; };
 
+  stlink-tool = callPackage ../development/tools/misc/stlink-tool { };
```
Comment (wegank) [inline pkgs/top-level/all-packages.nix:25392]: Please move the package to `pkgs/by-name/st/stlink-tool/package.nix` and drop this line, thanks.

File: pkgs/development/tools/misc/stlink-tool/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+# IMPORTANT: You need permissions to access the stlink usb devices.
+# Add services.udev.packages = [ pkgs.stlink ] to your configuration.nix
+stdenv.mkDerivation rec {
+  pname = "stlink-tool";
+  version = "unstable-2023-10-20";
+
+  src = fetchFromGitHub {
+    owner = "UweBonnes";
+    repo = pname;
```
Comment (wegank) [inline pkgs/development/tools/misc/stlink-tool/default.nix:16]:
```suggestion
    repo = "stlink-tool";
```

File: pkgs/development/tools/misc/stlink-tool/default.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    rev = "3ab0b140561c18c6e51183877c849b7b8ca0319c";
+    sha256 = "sha256-Wj72kHckxnox0kfqEZzE5zeSi8Bc+QSHPs4BtIf6RhE=";
```
Comment (wegank) [inline pkgs/development/tools/misc/stlink-tool/default.nix:18]:
```suggestion
    hash = "sha256-Wj72kHckxnox0kfqEZzE5zeSi8Bc+QSHPs4BtIf6RhE=";
```

File: pkgs/development/tools/misc/stlink-tool/default.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    fetchSubmodules = true;
+  };
+
+  buildInputs = [ libusb1 ];
+  nativeBuildInputs = [ pkg-config ];
+
+  preBuild = ''
+    NIX_CFLAGS_COMPILE="-Wno-uninitialized $NIX_CFLAGS_COMPILE"
+  '';
```
Comment (wegank) [inline pkgs/development/tools/misc/stlink-tool/default.nix:27]:
```suggestion
  env.NIX_CFLAGS_COMPILE = "-Wno-uninitialized";
```

File: pkgs/development/tools/misc/stlink-tool/default.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  installPhase = ''
+    runHook preInstall
+    install -D stlink-tool $out/bin/stlink-tool
+    runHook postInstall
+  '';
+
+  meta = with lib;
+    {
+      description = "libusb tool for flashing chinese ST-Link dongles";
```
Comment (wegank) [inline pkgs/development/tools/misc/stlink-tool/default.nix:37]: Any homepage?

File: pkgs/development/tools/misc/stlink-tool/default.nix (hunk 4 above)
Comment (wucke13) [reply pkgs/development/tools/misc/stlink-tool/default.nix:27]: Doesn't this override the flags?

Comment (wegank) [reply pkgs/development/tools/misc/stlink-tool/default.nix:27]: No.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (wucke13) [reply pkgs/top-level/all-packages.nix:25392]: Done

File: pkgs/development/tools/misc/stlink-tool/default.nix (hunk 5 above)
Comment (wucke13) [reply pkgs/development/tools/misc/stlink-tool/default.nix]: Unfortunately only the repo, adding it

File: pkgs/development/tools/misc/stlink-tool/default.nix (hunk 4 above)
Comment (wucke13) [reply pkgs/development/tools/misc/stlink-tool/default.nix]: Perfect, adding it

File: pkgs/development/tools/misc/stlink-tool/default.nix (hunk 3 above)
Comment (wucke13) [reply pkgs/development/tools/misc/stlink-tool/default.nix]: Rebased of the original (non-fork) repo, so obsolete now.

File: pkgs/development/tools/misc/stlink-tool/default.nix (hunk 2 above)
Comment (wucke13) [reply pkgs/development/tools/misc/stlink-tool/default.nix]: Done

**Comment by wucke13** [comment]
@wegank Thank you for the review, and sorry for the long delay. I adjusted (mostly) everything according to your suggestions and rebased.

**Comment by wucke13** [comment]
@GrahamcOfBorg build stlink-tool

File: pkgs/by-name/st/stlink-tool/package.nix [hunk 6]
```diff
@@ -0,0 +1,40 @@
+{ lib
+, stdenv
+, pkg-config
+, fetchFromGitHub
+, libusb1
+}:
+
+# IMPORTANT: You need permissions to access the stlink usb devices.
+# Add services.udev.packages = [ pkgs.stlink ] to your configuration.nix
+stdenv.mkDerivation rec {
```
Comment (wegank) [inline pkgs/by-name/st/stlink-tool/package.nix:10]:
```suggestion
stdenv.mkDerivation {
```

LGTM otherwise.

Comment (wucke13) [reply pkgs/by-name/st/stlink-tool/package.nix]: Done

**Review by wegank — APPROVED** [review]

**Comment by wucke13** [comment]
@GrahamcOfBorg build stlink-tool
@Lassulus Would you please consider to merge this?

**Pull request merged**